pkg/agent/plugin/nodeattestor/tpmdevid/tpmutil/signingkey.go

```diff
@@ -0,0 +1,180 @@
+		if rsaParams != nil {
+			sigHashAlg = rsaParams.Sign.Hash
+		}
+
+	case tpm2.AlgECC:
+		srkTemplate = tpm2tools.SRKTemplateECC()
+		eccParams := pub.ECCParameters
+		if eccParams != nil {
+			sigHashAlg = eccParams.Sign.Hash
+		}
+
+	default:
+		return nil, fmt.Errorf("bad key type: 0x%04x", pub.Type)
+	}
+
+	if sigHashAlg.IsNull() {
+		return nil, fmt.Errorf("signature hash algorithm is NULL")
+	}
+
+	srk, err := tpm2tools.NewKey(rw, tpm2.HandleOwner, srkTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("SRK creation failed: %w", err)
+	}
+
+	defer srk.Close()
+
+	keyHandle, _, err := tpm2.Load(rw, srk.Handle(), "", pubKey, privKey)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Load failed: %w", err)
+	}
+
+	return &SigningKey{
+		Handle:     keyHandle,
+		sigHashAlg: sigHashAlg,
+		rw:         rw,
+		log:        log,
+	}, nil
+}
+
+// Close removes the key from the TPM
+func (k *SigningKey) Close() error {
+	return tpm2.FlushContext(k.rw, k.Handle)
+}
+
+// Sign request the TPM to sign the given data using this key
+func (k *SigningKey) Sign(data []byte) ([]byte, error) {
+	digest, token, err := tpm2.Hash(k.rw, k.sigHashAlg, data, tpm2.HandlePlatform)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Hash failed: %w", err)
+	}
+
+	var sig *tpm2.Signature
+	for i := 0; i <= maxAttempts; i++ {
+		sig, err = tpm2.Sign(k.rw, k.Handle, "", digest, token, nil)
+		switch {
+		case err == nil:
+			break
+
+		case isRetry(err):
+			if i == maxAttempts {
+				return nil, fmt.Errorf("max attempts reached: %w", err)
+			}
+
+			k.log.Debug(fmt.Sprintf("TPM was not able to start the command 'Sign'. Retrying: attempt (%d/%d)", i, maxAttempts))
+			time.Sleep(time.Millisecond * 500)
+			continue
+
+		default:
+			return nil, fmt.Errorf("tpm2.Sign failed: %w", err)
+		}
+	}
+
+	if sig.RSA != nil {
+		return sig.RSA.Signature, nil
+	}
+
+	if sig.ECC != nil {
+		var b cryptobyte.Builder
+		b.AddASN1(asn1.SEQUENCE, func(b *cryptobyte.Builder) {
+			b.AddASN1BigInt(sig.ECC.R)
+			b.AddASN1BigInt(sig.ECC.S)
+		})
+
+		return b.Bytes()
+	}
+
+	return nil, fmt.Errorf("bad tpm2.Signature")
+}
+
+// Certify calls tpm2.Certify using the current key as signer and the provided
+// object as object handle.
+func (k *SigningKey) Certify(object tpmutil.Handle) ([]byte, []byte, error) {
+	// For some reason 'tpm2.Certify()' sometimes fails the first attempt and asks for retry.
+	// So, we retry some times in case of getting the RCRetry error.
+	// It seems that this issue has been reported: https://github.com/google/go-tpm/issues/59
+	var certifiedDevID []byte
+	var certificationSignature []byte
+	var err error
+	for i := 0; i <= maxAttempts; i++ {
+		certifiedDevID, certificationSignature, err = tpm2.Certify(k.rw, "", "", object, k.Handle, nil)
+		switch {
+		case err == nil:
+			return certifiedDevID, certificationSignature, nil
+
+		case isRetry(err):
+			if i == maxAttempts {
+				return nil, nil, fmt.Errorf("max attempts reached: %w", err)
```

Comment:
I think that here also would benefit from being more detailed about the action that is being performed while the error happened.